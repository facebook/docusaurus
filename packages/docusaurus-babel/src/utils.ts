/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import {BABEL_CONFIG_FILE_NAME} from '@docusaurus/utils';
import type {TransformOptions} from '@babel/core';

export async function getCustomBabelConfigFilePath(
  siteDir: string,
): Promise<string | undefined> {
  const customBabelConfigurationPath = path.join(
    siteDir,
    BABEL_CONFIG_FILE_NAME,
  );
  return (await fs.pathExists(customBabelConfigurationPath))
    ? customBabelConfigurationPath
    : undefined;
}

export function getBabelOptions({
  isServer,
  babelOptions,
}: {
  isServer?: boolean;
  // TODO Docusaurus v4 fix this
  //  weird to have getBabelOptions take a babelOptions param
  babelOptions?: TransformOptions | string;
} = {}): TransformOptions {
  const caller = {name: isServer ? 'server' : 'client'};
  if (typeof babelOptions === 'string') {
    return {
      babelrc: false,
      configFile: babelOptions,
      caller,
    };
  }
  return {
    ...(babelOptions ?? {
      presets: [require.resolve('@docusaurus/babel/preset')],
    }),
    babelrc: false,
    configFile: false,
    caller,
  };
}
