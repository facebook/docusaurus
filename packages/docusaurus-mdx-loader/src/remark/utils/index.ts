/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import escapeHtml from 'escape-html';
import type {Parent, Node} from 'unist';
import type {PhrasingContent, Heading} from 'mdast';
import type {
  MdxJsxAttribute,
  MdxJsxAttributeValueExpression,
  MdxJsxTextElement,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';

/**
 * Util to transform one node type to another node type
 * The input node is mutated in place
 * @param node the node to mutate
 * @param newNode what the original node should become become
 */
export function transformNode<NewNode extends Node>(
  node: Node,
  newNode: NewNode,
): NewNode {
  Object.keys(node).forEach((key) => {
    // @ts-expect-error: unsafe but ok
    delete node[key];
  });
  Object.keys(newNode).forEach((key) => {
    // @ts-expect-error: unsafe but ok
    node[key] = newNode[key];
  });
  return node as NewNode;
}

export function stringifyContent(
  node: Parent,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  return (node.children as PhrasingContent[])
    .map((item) => toValue(item, toString))
    .join('');
}

// TODO This is really a workaround, and not super reliable
// For now we only support serializing tagName, className and content
// Can we implement the TOC with real JSX nodes instead of html strings later?
function mdxJsxTextElementToHtml(
  element: MdxJsxTextElement,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  const tag = element.name;

  const attributes = element.attributes.filter(
    (child): child is MdxJsxAttribute => child.type === 'mdxJsxAttribute',
  );

  const classAttribute =
    attributes.find((attr) => attr.name === 'className') ??
    attributes.find((attr) => attr.name === 'class');

  const classAttributeString = classAttribute
    ? `class="${escapeHtml(String(classAttribute.value))}"`
    : ``;

  const allAttributes = classAttributeString ? ` ${classAttributeString}` : '';

  const content = stringifyContent(element, toString);

  return `<${tag}${allAttributes}>${content}</${tag}>`;
}

export function toValue(
  node: PhrasingContent | Heading | MdxJsxTextElement,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  switch (node.type) {
    case 'mdxJsxTextElement': {
      return mdxJsxTextElementToHtml(node as MdxJsxTextElement, toString);
    }
    case 'text':
      return escapeHtml(node.value);
    case 'heading':
      return stringifyContent(node, toString);
    case 'inlineCode':
      return `<code>${escapeHtml(node.value)}</code>`;
    case 'emphasis':
      return `<em>${stringifyContent(node, toString)}</em>`;
    case 'strong':
      return `<strong>${stringifyContent(node, toString)}</strong>`;
    case 'delete':
      return `<del>${stringifyContent(node, toString)}</del>`;
    case 'link':
      return stringifyContent(node, toString);
    default:
      return toString(node);
  }
}

export function assetRequireAttributeValue(
  requireString: string,
  hash: string,
): MdxJsxAttributeValueExpression {
  return {
    type: 'mdxJsxAttributeValueExpression',
    value: `require("${requireString}").default${hash && ` + '${hash}'`}`,
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'BinaryExpression',
              left: {
                type: 'MemberExpression',
                object: {
                  type: 'CallExpression',
                  callee: {
                    type: 'Identifier',
                    name: 'require',
                  },
                  arguments: [
                    {
                      type: 'Literal',
                      value: requireString,
                      raw: `"${requireString}"`,
                    },
                  ],
                  optional: false,
                },
                property: {
                  type: 'Identifier',
                  name: 'default',
                },
                computed: false,
                optional: false,
              },
              operator: '+',
              right: {
                type: 'Literal',
                value: hash,
                raw: `"${hash}"`,
              },
            },
          },
        ],
        sourceType: 'module',
        comments: [],
      },
    },
  };
}
