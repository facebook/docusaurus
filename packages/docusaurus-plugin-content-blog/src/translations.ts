/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {TranslationFileContent, TranslationFile} from '@docusaurus/types';
import type {
  PluginOptions,
  BlogContent,
  BlogPaginated,
} from '@docusaurus/plugin-content-blog';

function translateListPage(
  blogListPaginated: BlogPaginated[],
  translations: TranslationFileContent,
) {
  return blogListPaginated.map((page) => {
    const {items, metadata} = page;
    return {
      items,
      metadata: {
        ...metadata,
        blogTitle: translations.title?.message ?? page.metadata.blogTitle,
        blogDescription:
          translations.description?.message ?? page.metadata.blogDescription,
      },
    };
  });
}

export function getTranslationFiles(options: PluginOptions): TranslationFile[] {
  return [
    {
      path: 'options',
      content: {
        title: {
          message: options.blogTitle,
          description: 'The title for the blog used in SEO',
        },
        description: {
          message: options.blogDescription,
          description: 'The description for the blog used in SEO',
        },
        'sidebar.title': {
          message: options.blogSidebarTitle,
          description: 'The label for the left sidebar',
        },
      },
    },
  ];
}

export function translateContent(
  content: BlogContent,
  translationFiles: TranslationFile[],
): BlogContent {
  const {content: optionsTranslations} = translationFiles[0]!;
  return {
    ...content,
    blogSidebarTitle:
      optionsTranslations['sidebar.title']?.message ?? content.blogSidebarTitle,
    blogListPaginated: translateListPage(
      content.blogListPaginated,
      optionsTranslations,
    ),
  };
}
