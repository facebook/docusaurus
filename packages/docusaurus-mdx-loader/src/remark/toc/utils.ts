/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {toValue} from '../utils';
import type {Node} from 'unist';
import type {
  MdxjsEsm,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';
import type {TOCHeading, TOCItem, TOCItems, TOCSlice} from './types';
import type {
  Program,
  SpreadElement,
  ImportDeclaration,
  ImportSpecifier,
} from 'estree';

export function getImportDeclarations(program: Program): ImportDeclaration[] {
  return program.body.filter(
    (item): item is ImportDeclaration => item.type === 'ImportDeclaration',
  );
}

export function isMarkdownImport(node: Node): node is ImportDeclaration {
  if (node.type !== 'ImportDeclaration') {
    return false;
  }
  const importPath = (node as ImportDeclaration).source.value;
  return typeof importPath === 'string' && /\.mdx?$/.test(importPath);
}

export function findDefaultImportName(
  importDeclaration: ImportDeclaration,
): string | undefined {
  return importDeclaration.specifiers.find(
    (o: Node) => o.type === 'ImportDefaultSpecifier',
  )?.local.name;
}

export function findNamedImportSpecifier(
  importDeclaration: ImportDeclaration,
  localName: string,
): ImportSpecifier | undefined {
  return importDeclaration?.specifiers.find(
    (specifier): specifier is ImportSpecifier =>
      specifier.type === 'ImportSpecifier' &&
      specifier.local.name === localName,
  );
}

// Before: import Partial from "partial"
// After: import Partial, {toc as __tocPartial} from "partial"
export function addTocSliceImportIfNeeded({
  importDeclaration,
  tocExportName,
  tocSliceImportName,
}: {
  importDeclaration: ImportDeclaration;
  tocExportName: string;
  tocSliceImportName: string;
}): void {
  // We only add the toc slice named import if it doesn't exist already
  if (!findNamedImportSpecifier(importDeclaration, tocSliceImportName)) {
    importDeclaration.specifiers.push({
      type: 'ImportSpecifier',
      imported: {type: 'Identifier', name: tocExportName},
      local: {type: 'Identifier', name: tocSliceImportName},
    });
  }
}

export function isNamedExport(
  node: Node,
  exportName: string,
): node is MdxjsEsm {
  if (node.type !== 'mdxjsEsm') {
    return false;
  }
  const program = (node as MdxjsEsm).data?.estree;
  if (!program) {
    return false;
  }
  if (program.body.length !== 1) {
    return false;
  }
  const exportDeclaration = program.body[0]!;
  if (exportDeclaration.type !== 'ExportNamedDeclaration') {
    return false;
  }
  const variableDeclaration = exportDeclaration.declaration;
  if (variableDeclaration?.type !== 'VariableDeclaration') {
    return false;
  }
  const {id} = variableDeclaration.declarations[0]!;
  if (id.type !== 'Identifier') {
    return false;
  }

  return id.name === exportName;
}

export async function createTOCExportNodeAST({
  tocExportName,
  tocItems,
}: {
  tocExportName: string;
  tocItems: TOCItems;
}): Promise<MdxjsEsm> {
  function createTOCSliceAST(tocSlice: TOCSlice): SpreadElement {
    return {
      type: 'SpreadElement',
      argument: {type: 'Identifier', name: tocSlice.importName},
    };
  }

  async function createTOCHeadingAST({heading}: TOCHeading) {
    const {toString} = await import('mdast-util-to-string');
    const {valueToEstree} = await import('estree-util-value-to-estree');
    const value: TOCItem = {
      value: toValue(heading, toString),
      id: heading.data!.id!,
      level: heading.depth,
    };
    return valueToEstree(value);
  }

  async function createTOCItemAST(tocItem: TOCItems[number]) {
    switch (tocItem.type) {
      case 'slice':
        return createTOCSliceAST(tocItem);
      case 'heading':
        return createTOCHeadingAST(tocItem);
      default: {
        throw new Error(`unexpected toc item type`);
      }
    }
  }

  return {
    type: 'mdxjsEsm',
    value: '', // See https://github.com/facebook/docusaurus/pull/9684#discussion_r1457595181
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ExportNamedDeclaration',
            declaration: {
              type: 'VariableDeclaration',
              declarations: [
                {
                  type: 'VariableDeclarator',
                  id: {
                    type: 'Identifier',
                    name: tocExportName,
                  },
                  init: {
                    type: 'ArrayExpression',
                    elements: await Promise.all(tocItems.map(createTOCItemAST)),
                  },
                },
              ],
              kind: 'const',
            },
            specifiers: [],
            source: null,
          },
        ],
        sourceType: 'module',
      },
    },
  };
}
