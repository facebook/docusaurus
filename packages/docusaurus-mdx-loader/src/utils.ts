/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import logger from '@docusaurus/logger';
import {escapePath, type WebpackCompilerName} from '@docusaurus/utils';
import {getProcessor, type SimpleProcessorResult} from './processor';
import {validateMDXFrontMatter} from './frontMatter';
import preprocessor from './preprocessor';
import type {Options} from './loader';

/**
 * When this throws, it generally means that there's no metadata file associated
 * with this MDX document. It can happen when using MDX partials (usually
 * starting with _). That's why it's important to provide the `isMDXPartial`
 * function in config
 */
export async function readMetadataPath(metadataPath: string): Promise<unknown> {
  try {
    return await fs.readJSON(metadataPath, 'utf8');
  } catch (error) {
    throw new Error(
      logger.interpolate`MDX loader can't read MDX metadata file path=${metadataPath}. Maybe the isMDXPartial option function was not provided?`,
      {cause: error as Error},
    );
  }
}

/**
 * Converts assets an object with Webpack require calls code.
 * This is useful for mdx files to reference co-located assets using relative
 * paths. Those assets should enter the Webpack assets pipeline and be hashed.
 * For now, we only handle that for images and paths starting with `./`:
 *
 * `{image: "./myImage.png"}` => `{image: require("./myImage.png")}`
 */
export function createAssetsExportCode({
  assets,
  inlineMarkdownAssetImageFileLoader,
}: {
  assets: unknown;
  inlineMarkdownAssetImageFileLoader: string;
}): string {
  if (
    typeof assets !== 'object' ||
    !assets ||
    Object.keys(assets).length === 0
  ) {
    return 'undefined';
  }

  // TODO implementation can be completed/enhanced
  function createAssetValueCode(assetValue: unknown): string | undefined {
    if (Array.isArray(assetValue)) {
      const arrayItemCodes = assetValue.map(
        (item: unknown) => createAssetValueCode(item) ?? 'undefined',
      );
      return `[${arrayItemCodes.join(', ')}]`;
    }
    // Only process string values starting with ./
    // We could enhance this logic and check if file exists on disc?
    if (typeof assetValue === 'string' && assetValue.startsWith('./')) {
      // TODO do we have other use-cases than image assets?
      // Probably not worth adding more support, as we want to move to Webpack 5 new asset system (https://github.com/facebook/docusaurus/pull/4708)
      return `require("${inlineMarkdownAssetImageFileLoader}${escapePath(
        assetValue,
      )}").default`;
    }
    return undefined;
  }

  const assetEntries = Object.entries(assets);

  const codeLines = assetEntries
    .map(([key, value]: [string, unknown]) => {
      const assetRequireCode = createAssetValueCode(value);
      return assetRequireCode ? `"${key}": ${assetRequireCode},` : undefined;
    })
    .filter(Boolean);

  return `{\n${codeLines.join('\n')}\n}`;
}

/**
 * data.contentTitle is set by the remark contentTitle plugin
 */
export function extractContentTitleData(data: {
  [key: string]: unknown;
}): string | undefined {
  return data.contentTitle as string | undefined;
}

export async function compileToJSX({
  filePath,
  fileContent,
  frontMatter,
  options,
  compilerName,
}: {
  filePath: string;
  fileContent: string;
  frontMatter: Record<string, unknown>;
  options: Options;
  compilerName: WebpackCompilerName;
}): Promise<SimpleProcessorResult> {
  const preprocessedFileContent = preprocessor({
    fileContent,
    filePath,
    admonitions: options.admonitions,
    markdownConfig: options.markdownConfig,
  });

  const mdxFrontMatter = validateMDXFrontMatter(frontMatter.mdx);

  const processor = await getProcessor({
    filePath,
    options,
    mdxFrontMatter,
  });

  try {
    return await processor.process({
      content: preprocessedFileContent,
      filePath,
      frontMatter,
      compilerName,
    });
  } catch (errorUnknown) {
    const error = errorUnknown as Error;

    // MDX can emit errors that have useful extra attributes
    const errorJSON = JSON.stringify(error, null, 2);
    const errorDetails =
      errorJSON === '{}'
        ? // regular JS error case: print stacktrace
          error.stack ?? 'N/A'
        : // MDX error: print extra attributes + stacktrace
          `${errorJSON}\n${error.stack}`;

    throw new Error(
      `MDX compilation failed for file ${logger.path(filePath)}\nCause: ${
        error.message
      }\nDetails:\n${errorDetails}`,
      // TODO error cause doesn't seem to be used by Webpack stats.errors :s
      {cause: error},
    );
  }
}
