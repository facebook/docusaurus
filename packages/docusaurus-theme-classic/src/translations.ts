/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import _ from 'lodash';
import {mergeTranslations} from '@docusaurus/utils';
import type {TranslationFile, TranslationFileContent} from '@docusaurus/types';
import type {
  ThemeConfig,
  Navbar,
  NavbarItem,
  Footer,
  MultiColumnFooter,
  SimpleFooter,
} from '@docusaurus/theme-common';

function getNavbarTranslationFile(navbar: Navbar): TranslationFileContent {
  // TODO handle properly all the navbar item types here!
  function flattenNavbarItems(items: NavbarItem[]): NavbarItem[] {
    const subItems = items.flatMap((item) => {
      const allSubItems = [item.items ?? []].flat();
      return flattenNavbarItems(allSubItems);
    });
    return [...items, ...subItems];
  }

  const allNavbarItems = flattenNavbarItems(navbar.items);

  const navbarItemsTranslations: TranslationFileContent = Object.fromEntries(
    allNavbarItems
      .filter((navbarItem) => navbarItem.label)
      .map((navbarItem) => [
        `item.label.${navbarItem.label}`,
        {
          message: navbarItem.label!,
          description: `Navbar item with label ${navbarItem.label}`,
        },
      ]),
  );

  const titleTranslations: TranslationFileContent = navbar.title
    ? {title: {message: navbar.title, description: 'The title in the navbar'}}
    : {};

  const logoAlt: TranslationFileContent = navbar.logo?.alt
    ? {
        'logo.alt': {
          message: navbar.logo.alt,
          description: 'The alt text of navbar logo',
        },
      }
    : {};

  return mergeTranslations([
    titleTranslations,
    logoAlt,
    navbarItemsTranslations,
  ]);
}
function translateNavbar(
  navbar: Navbar,
  navbarTranslations: TranslationFileContent | undefined,
): Navbar {
  if (!navbarTranslations) {
    return navbar;
  }

  const logo = navbar.logo
    ? {
        ...navbar.logo,
        alt: navbarTranslations[`logo.alt`]?.message ?? navbar.logo?.alt,
      }
    : undefined;

  return {
    ...navbar,
    title: navbarTranslations.title?.message ?? navbar.title,
    logo,
    //  TODO handle properly all the navbar item types here!
    items: navbar.items.map((item) => {
      const subItems = item.items?.map((subItem) => ({
        ...subItem,
        label:
          navbarTranslations[`item.label.${subItem.label}`]?.message ??
          subItem.label,
      }));
      return {
        ...item,
        label:
          navbarTranslations[`item.label.${item.label}`]?.message ?? item.label,
        ...(subItems ? {items: subItems} : undefined),
      };
    }),
  };
}

function isMultiColumnFooterLinks(
  links: MultiColumnFooter['links'] | SimpleFooter['links'],
): links is MultiColumnFooter['links'] {
  return links.length > 0 && 'title' in links[0]!;
}

function getFooterTranslationFile(footer: Footer): TranslationFileContent {
  const footerLinkTitles: TranslationFileContent = Object.fromEntries(
    (isMultiColumnFooterLinks(footer.links)
      ? footer.links.filter((link) => link.title)
      : []
    ).map((link) => [
      `link.title.${link.title}`,
      {
        message: link.title!,
        description: `The title of the footer links column with title=${link.title} in the footer`,
      },
    ]),
  );

  const footerLinkLabels: TranslationFileContent = Object.fromEntries(
    (isMultiColumnFooterLinks(footer.links)
      ? footer.links.flatMap((link) => link.items).filter((link) => link.label)
      : footer.links.filter((link) => link.label)
    ).map((link) => [
      `link.item.label.${link.label}`,
      {
        message: link.label!,
        description: `The label of footer link with label=${
          link.label
        } linking to ${link.to ?? link.href}`,
      },
    ]),
  );

  const copyright: TranslationFileContent = footer.copyright
    ? {
        copyright: {
          message: footer.copyright,
          description: 'The footer copyright',
        },
      }
    : {};

  const logoAlt: TranslationFileContent = footer.logo?.alt
    ? {
        'logo.alt': {
          message: footer.logo.alt,
          description: 'The alt text of footer logo',
        },
      }
    : {};

  return mergeTranslations([
    footerLinkTitles,
    footerLinkLabels,
    copyright,
    logoAlt,
  ]);
}
function translateFooter(
  footer: Footer,
  footerTranslations: TranslationFileContent | undefined,
): Footer {
  if (!footerTranslations) {
    return footer;
  }
  const links = isMultiColumnFooterLinks(footer.links)
    ? footer.links.map((link) => ({
        ...link,
        title:
          footerTranslations[`link.title.${link.title}`]?.message ?? link.title,
        items: link.items.map((linkItem) => ({
          ...linkItem,
          label:
            footerTranslations[`link.item.label.${linkItem.label}`]?.message ??
            linkItem.label,
        })),
      }))
    : footer.links.map((link) => ({
        ...link,
        label:
          footerTranslations[`link.item.label.${link.label}`]?.message ??
          link.label,
      }));

  const copyright = footerTranslations.copyright?.message ?? footer.copyright;

  const logo = footer.logo
    ? {
        ...footer.logo,
        alt: footerTranslations[`logo.alt`]?.message ?? footer.logo?.alt,
      }
    : undefined;

  return {
    ...footer,
    links,
    copyright,
    logo,
  };
}

export function getTranslationFiles({
  themeConfig,
}: {
  themeConfig: ThemeConfig;
}): TranslationFile[] {
  const translationFiles: (TranslationFile | undefined)[] = [
    {path: 'navbar', content: getNavbarTranslationFile(themeConfig.navbar)},
    themeConfig.footer
      ? {
          path: 'footer',
          content: getFooterTranslationFile(themeConfig.footer),
        }
      : undefined,
  ];

  return translationFiles.filter(Boolean) as TranslationFile[];
}

export function translateThemeConfig({
  themeConfig,
  translationFiles,
}: {
  // To make TS correctly figure out the contravariance in parameter.
  // In practice it's always normalized
  themeConfig: ThemeConfig;
  translationFiles: TranslationFile[];
}): ThemeConfig {
  const translationFilesMap: {[fileName: string]: TranslationFile} = _.keyBy(
    translationFiles,
    (f) => f.path,
  );

  return {
    ...themeConfig,
    navbar: translateNavbar(
      themeConfig.navbar,
      translationFilesMap.navbar?.content,
    ),
    footer: themeConfig.footer
      ? translateFooter(themeConfig.footer, translationFilesMap.footer?.content)
      : undefined,
  };
}
