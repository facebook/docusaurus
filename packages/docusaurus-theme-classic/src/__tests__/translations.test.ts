/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {updateTranslationFileMessages} from '@docusaurus/utils';
import {getTranslationFiles, translateThemeConfig} from '../translations';
import type {ThemeConfig} from '@docusaurus/theme-common';

const ThemeConfigSample = {
  colorMode: {},
  announcementBar: {},
  prism: {},
  docs: {
    versionPersistence: 'none',
  },
  navbar: {
    title: 'navbar title',
    logo: {
      alt: 'navbar alt text logo',
      src: 'img/docusaurus.svg',
    },
    style: 'dark',
    hideOnScroll: false,
    items: [
      {
        label: 'Dropdown',
        items: [
          {label: 'Dropdown item 1', items: []},
          {label: 'Dropdown item 2', items: []},
        ],
      },
    ],
  },
  footer: {
    logo: {
      alt: 'footer alt text logo',
      src: 'img/docusaurus.svg',
    },
    copyright: 'Copyright FB',
    style: 'light',
    links: [
      {
        title: 'Footer link column 1',
        items: [
          {label: 'Link 1', to: 'https://facebook.com'},
          {label: 'Link 2', to: 'https://facebook.com'},
        ],
      },
      {
        title: 'Footer link column 2',
        items: [{label: 'Link 3', to: 'https://facebook.com'}],
      },
    ],
  },
} as unknown as ThemeConfig;

const ThemeConfigSampleSimpleFooter: ThemeConfig = {
  ...ThemeConfigSample,
  footer: {
    logo: {
      alt: 'footer alt text logo',
      src: 'img/docusaurus.svg',
    },
    copyright: 'Copyright FB',
    style: 'light',
    links: [
      {label: 'Link 1', to: 'https://facebook.com'},
      {label: 'Link 2', to: 'https://facebook.com'},
    ],
  },
};

function getSampleTranslationFiles(themeConfig: ThemeConfig) {
  return getTranslationFiles({
    themeConfig,
  });
}

function getSampleTranslationFilesTranslated(themeConfig: ThemeConfig) {
  const translationFiles = getSampleTranslationFiles(themeConfig);
  return translationFiles.map((translationFile) =>
    updateTranslationFileMessages(
      translationFile,
      (message) => `${message} (translated)`,
    ),
  );
}

describe('getTranslationFiles', () => {
  it('returns translation files matching snapshot', () => {
    expect(getSampleTranslationFiles(ThemeConfigSample)).toMatchSnapshot();
    expect(
      getSampleTranslationFiles(ThemeConfigSampleSimpleFooter),
    ).toMatchSnapshot();
  });
});

describe('translateThemeConfig', () => {
  it('does not translate anything if translation files are untranslated', () => {
    expect(
      translateThemeConfig({
        themeConfig: ThemeConfigSample,
        translationFiles: getSampleTranslationFiles(ThemeConfigSample),
      }),
    ).toEqual(ThemeConfigSample);
  });

  it('returns translated themeConfig', () => {
    expect(
      translateThemeConfig({
        themeConfig: ThemeConfigSample,
        translationFiles:
          getSampleTranslationFilesTranslated(ThemeConfigSample),
      }),
    ).toMatchSnapshot();
  });
});

describe('getTranslationFiles and translateThemeConfig isomorphism', () => {
  function verifyIsomorphism(themeConfig: ThemeConfig) {
    const translatedThemeConfig = translateThemeConfig({
      themeConfig,
      translationFiles: getTranslationFiles({themeConfig}),
    });
    expect(translatedThemeConfig).toEqual(themeConfig);
  }

  it('is verified for sample', () => {
    verifyIsomorphism(ThemeConfigSample);
  });

  it('is verified for sample with simple footer', () => {
    verifyIsomorphism(ThemeConfigSampleSimpleFooter);
  });

  // Undefined footer should not make the translation code crash
  // See https://github.com/facebook/docusaurus/issues/3936
  it('is verified for sample without footer', () => {
    verifyIsomorphism({...ThemeConfigSample, footer: undefined});
  });
});
