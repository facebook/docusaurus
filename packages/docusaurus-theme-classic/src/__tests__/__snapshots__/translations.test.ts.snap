// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTranslationFiles returns translation files matching snapshot 1`] = `
[
  {
    "content": {
      "item.label.Dropdown": {
        "description": "Navbar item with label Dropdown",
        "message": "Dropdown",
      },
      "item.label.Dropdown item 1": {
        "description": "Navbar item with label Dropdown item 1",
        "message": "Dropdown item 1",
      },
      "item.label.Dropdown item 2": {
        "description": "Navbar item with label Dropdown item 2",
        "message": "Dropdown item 2",
      },
      "logo.alt": {
        "description": "The alt text of navbar logo",
        "message": "navbar alt text logo",
      },
      "title": {
        "description": "The title in the navbar",
        "message": "navbar title",
      },
    },
    "path": "navbar",
  },
  {
    "content": {
      "copyright": {
        "description": "The footer copyright",
        "message": "Copyright FB",
      },
      "link.item.label.Link 1": {
        "description": "The label of footer link with label=Link 1 linking to https://facebook.com",
        "message": "Link 1",
      },
      "link.item.label.Link 2": {
        "description": "The label of footer link with label=Link 2 linking to https://facebook.com",
        "message": "Link 2",
      },
      "link.item.label.Link 3": {
        "description": "The label of footer link with label=Link 3 linking to https://facebook.com",
        "message": "Link 3",
      },
      "link.title.Footer link column 1": {
        "description": "The title of the footer links column with title=Footer link column 1 in the footer",
        "message": "Footer link column 1",
      },
      "link.title.Footer link column 2": {
        "description": "The title of the footer links column with title=Footer link column 2 in the footer",
        "message": "Footer link column 2",
      },
      "logo.alt": {
        "description": "The alt text of footer logo",
        "message": "footer alt text logo",
      },
    },
    "path": "footer",
  },
]
`;

exports[`getTranslationFiles returns translation files matching snapshot 2`] = `
[
  {
    "content": {
      "item.label.Dropdown": {
        "description": "Navbar item with label Dropdown",
        "message": "Dropdown",
      },
      "item.label.Dropdown item 1": {
        "description": "Navbar item with label Dropdown item 1",
        "message": "Dropdown item 1",
      },
      "item.label.Dropdown item 2": {
        "description": "Navbar item with label Dropdown item 2",
        "message": "Dropdown item 2",
      },
      "logo.alt": {
        "description": "The alt text of navbar logo",
        "message": "navbar alt text logo",
      },
      "title": {
        "description": "The title in the navbar",
        "message": "navbar title",
      },
    },
    "path": "navbar",
  },
  {
    "content": {
      "copyright": {
        "description": "The footer copyright",
        "message": "Copyright FB",
      },
      "link.item.label.Link 1": {
        "description": "The label of footer link with label=Link 1 linking to https://facebook.com",
        "message": "Link 1",
      },
      "link.item.label.Link 2": {
        "description": "The label of footer link with label=Link 2 linking to https://facebook.com",
        "message": "Link 2",
      },
      "logo.alt": {
        "description": "The alt text of footer logo",
        "message": "footer alt text logo",
      },
    },
    "path": "footer",
  },
]
`;

exports[`translateThemeConfig returns translated themeConfig 1`] = `
{
  "announcementBar": {},
  "colorMode": {},
  "docs": {
    "versionPersistence": "none",
  },
  "footer": {
    "copyright": "Copyright FB (translated)",
    "links": [
      {
        "items": [
          {
            "label": "Link 1 (translated)",
            "to": "https://facebook.com",
          },
          {
            "label": "Link 2 (translated)",
            "to": "https://facebook.com",
          },
        ],
        "title": "Footer link column 1 (translated)",
      },
      {
        "items": [
          {
            "label": "Link 3 (translated)",
            "to": "https://facebook.com",
          },
        ],
        "title": "Footer link column 2 (translated)",
      },
    ],
    "logo": {
      "alt": "footer alt text logo (translated)",
      "src": "img/docusaurus.svg",
    },
    "style": "light",
  },
  "navbar": {
    "hideOnScroll": false,
    "items": [
      {
        "items": [
          {
            "items": [],
            "label": "Dropdown item 1 (translated)",
          },
          {
            "items": [],
            "label": "Dropdown item 2 (translated)",
          },
        ],
        "label": "Dropdown (translated)",
      },
    ],
    "logo": {
      "alt": "navbar alt text logo (translated)",
      "src": "img/docusaurus.svg",
    },
    "style": "dark",
    "title": "navbar title (translated)",
  },
  "prism": {},
}
`;
